### PR TITLE
Load `ncdiag` module instead of building it in the GSI, use UFSWM build system, pull sfcanl out from anal.

### LINKED ISSUE #906: Update ncio version in GSI utils.
Update `ncio` module version in GSI-utils. 
This likely affects cycled after PR that split the GSI utilities into its own repo.

**Expected behavior**
Likely affecting ensemble post processing jobs. 

**Current behavior**
All post processing jobs should succeed but are likely failing. 

**Machines affected**
All

### PR DESCRIPTION
**Pull request by aerorahul**
**Description**

`ncdiag` is no longer being built within the GSI.  It is available as a module on all systems; Hera, Orion and WCOSS2.

This PR:
- loads ncdiag module
- finds the `ncdiag_cat_serial.x` executable from that module (once loaded)
- updates hashes for GSI, GSI-utils and GSI-monitor.  No code changes were made in these repos.  Only location of `ncdiag` changed.

In addition, this PR:
- Uses the ufs-weather-model compilation script.  The global workflow is duplicating and missing configuration options set in the ufs-weather-model.  This also enables a "debug" mode of compiling the model (#300).
- Fixes #906 
- Fixes an improper link to `fix_gsi`.  It was being linked to a non-existing directory.
- separates out the creation of Gaussian surface analysis from the GSI atmospheric analysis.  This capability is pulled from PR #871.  It was necessary to pull these changes from PR #871 since the `j-jobs`, `ex-scripts` from the GSI have been brought to global-workflow and the companion PR for PR #871 in the [GSI PR 415](https://github.com/NOAA-EMC/GSI/pull/415) were already merged prior to the updates needed for `ncio` and `ncdiag`.

I realize all these changes could have been streamlined into smaller atomic PR's.

**Note:**  This PR also requires a documentation update as a new job `sfcanl` has been added to the workflow.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [x] Ran a full 3DVar cycled system on Orion at C96 for 5.5 cycles, including 2 00Z `gfs` cycle.  All jobs ran to completion with the exception of `gfsmetpg2o1`. 

```
EXPDIR: /work2/noaa/stmp/rmahajan/work/EXPDIR/pr905b  
ROTDIR: /work2/noaa/stmp/rmahajan/work/ROTDIR/pr905b
```
  
**Checklist**

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [X] Any dependent changes have been merged and published
- [ ] Requires a documentation update.  New job `sfcanl` has been introduced in this PR.

### CODE REVIEW

File: sorc/build_ufs.sh [hunk 1]
```diff
@@ -1,44 +1,38 @@
...
 set -eux
 
+cwd=$(pwd)
+
 # Default settings
 APP="S2SWA"
 CCPP_SUITES="FV3_GFS_v16,FV3_GFS_v16_ugwpv1,FV3_GFS_v17_p8,FV3_GFS_v16_coupled_nsstNoahmpUGWPv1,FV3_GFS_v17_coupled_p8"
 
-while getopts "a:v" option; do
+while getopts ":dav" option; do
```
Comment (WalterKolczynski-NOAA) [inline sorc/build_ufs.sh]: The app argument will no longer work with this construction. Option string should be `:da:v`

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
I thought we were supposed to be moving away from build script to using cmake directly?

Otherwise, just one bug found upon reading the code.

File: sorc/build_ufs.sh (hunk 1 above)
Comment (aerorahul) [reply sorc/build_ufs.sh]: fixed.

**Comment by aerorahul** [comment]
> I thought we were supposed to be moving away from build script to using cmake directly?
> 
> Otherwise, just one bug found upon reading the code.

Re. building the component executables.
The components know what modules to load, what options to set, etc.  They use the build mechanism of their choice (CMake, etc).  To duplicate those would mean repeating those instructions.  Further, it also means explaining to the component owner how is "our" build different than the one they support on all platforms.
Along the principle of separation of concerns, we should (IMO), leverage as many as possible options from the components.  If @junwang-noaa feels that we should maintain our own build commands for the ufs-weather-model, I am happy to revert.

**Comment by WalterKolczynski-NOAA** [comment]
> > I thought we were supposed to be moving away from build script to using cmake directly?
> > Otherwise, just one bug found upon reading the code.
> 
> Re. building the component executables. The components know what modules to load, what options to set, etc. They use the build mechanism of their choice (CMake, etc). To duplicate those would mean repeating those instructions. Further, it also means explaining to the component owner how is "our" build different than the one they support on all platforms. Along the principle of separation of concerns, we should (IMO), leverage as many as possible options from the components. If @junwang-noaa feels that we should maintain our own build commands for the ufs-weather-model, I am happy to revert.

Yeah, I have no problem with it, I just recall an earlier discussion about not doing that. I'm happy to leverage other repo's scripts where we can.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Code approved pending a cycled test.

**Review by JessicaMeixner-NOAA — APPROVED** [review]
The added functionality to have debug with the model compile is a really appreciated and nice feature.  I tested and can confirm we can run a S2SW with debug on orion, which a great new capability. I cannot comment on the gsi related portions of this code.

**Review by WalterKolczynski-NOAA — APPROVED** [review]
Code still looks great, merge pending testing.

**Comment by aerorahul** [comment]
The testing on Orion has concluded.  Please see PR description.
A snapshot of the last 00z cycle is attached.
Screen Shot 2022-07-15 at 9 31 19 AM

**Review by RussTreadon-NOAA — APPROVED** [review]
Just returned from leave.  So as to not delay this PR, assume sufficient testing by @aerorahul and reviews by @WalterKolczynski-NOAA and @JessicaMeixner-NOAA.    Approve.

**Comment by RussTreadon-NOAA** [comment]
Examination of log files in `/work2/noaa/stmp/rmahajan/work/ROTDIR/pr905b/logs` and from my tests of this PR reveal a silent failure in `gdassfcanl`.

```
0.120 + [ 2021122100 -le ]
/work/noaa/stmp/rmahajan/work/global-workflow/scripts/exglobal_atmos_sfcanl.sh[165]: [: argument expected
0.120 + SNOW_NUDGE_COEFF=-2.
0.120 + export SNOW_NUDGE_COEFF
0.120 + CYCLVARS=FSNOL=-2.,
0.120 + export CYCLVARS
```
The right hand side of the `-le` test is a check of the date in file `snoprv`.   If we use `2021122200` as an example, the `gdassfcanl` job incorrectly sets `snoprv=/work/noaa/rstprod/dump/gdas.20211222/00/atmos/gdas.t18z.snogrb_t1534.3072.1536`.   File `gdas.t18z.snogrb_t1534.3072.1536` does not exist in path `/work/noaa/rstprod/dump/gdas.20211222/00/atmos`.   Variable `COMIN_GES_OBS` is incorrectly set to `/work/noaa/rstprod/dump/gdas.20211222/00/atmos`

Due to this error, `global_cycle` is run with `CYCLVARS=FSNOL=-2.,`.   The correct setting is `CYCLVARS=FSNOL=99999.,FSNOS=99999.,`

As a test, add the following to `config.base`
```
GDATE=$($NDATE -$assim_freq $CDATE)
gPDY=$(echo $GDATE | cut -c1-8)
gcyc=$(echo $GDATE | cut -c9-10)
GDUMP=${GDUMP:-"gdas"}
export COMIN_GES_OBS=${DMPDIR}/${GDUMP}.${gPDY}/$gcyc/atmos
```

Rerun gdassfcanl.   With this change the correct path is set for `snoprv` and we wind up with

```
0 + snoprv=/work/noaa/rstprod/dump/gdas.20211220/18/atmos/gdas.t18z.snogrb_t1534.3072.1536
...
0 + '[' 2021122000 -le 2021122000 ']'
0 + export 'FNSNOA= '
0 + FNSNOA=' '
0 + export CYCLVARS=FSNOL=99999.,FSNOS=99999.,
0 + CYCLVARS=FSNOL=99999.,FSNOS=99999.,
```

`config.base.emc.dyn` incorrectly sets `COMIN_GES_OBS` to the same value as `COMIN_OBS`
```
export COMIN_OBS=${DMPDIR}/${CDUMP}.${PDY}/$cyc/atmos
export COMIN_GES_OBS=${DMPDIR}/${CDUMP}.${PDY}/$cyc/atmos
```
If `ROTDIR_DUMP=YES`, as is the case in the above tests,  `JGLOBAL_ATMOS_ANALYSIS` uses the `config.base` setting for `COMIN_GES_OBS`.    (As an aside, I don't think `compath.py` works on Orion.)
```
if [ $RUN_ENVIR = "nco" -o ${ROTDIR_DUMP:-NO} = "YES" ]; then
    export COMIN=${COMIN:-$ROTDIR/$RUN.$PDY/$cyc/$COMPONENT}
    export COMOUT=${COMOUT:-$ROTDIR/$RUN.$PDY/$cyc/$COMPONENT}
    export COMIN_OBS=${COMIN_OBS:-$(compath.py ${envir}/obsproc/${obsproc_ver})/$RUN.$PDY/$cyc/$COMPONENT}
    export COMIN_GES_OBS=${COMIN_GES_OBS:-$(compath.py ${envir}/obsproc/${obsproc_ver})/$GDUMP.$gPDY/$gcyc/$COMPONENT}
else
    export COMOUT="$ROTDIR/$CDUMP.$PDY/$cyc/$COMPONENT"
    export COMIN_OBS="$DMPDIR/$CDUMP.$PDY/$cyc/$COMPONENT"
    export COMIN_GES_OBS="$DMPDIR/$GDUMP.$gPDY/$gcyc/$COMPONENT"
fi
```

Interestingly, `JGLOBAL_ATMOS_SFCANL` has different defaults for `COMIN_GES_OBS`
```
if [ $RUN_ENVIR = "nco" -o ${ROTDIR_DUMP:-NO} = "YES" ]; then
    export COMIN=${COMIN:-$ROTDIR/$RUN.$PDY/$cyc/$COMPONENT}
    export COMOUT=${COMOUT:-$ROTDIR/$RUN.$PDY/$cyc/$COMPONENT}
    export COMIN_OBS=${COMIN_OBS:-$ROTDIR/$RUN.$PDY/$cyc/$COMPONENT}
    export COMIN_GES_OBS=${COMIN_GES_OBS:-$ROTDIR/$GDUMP.$gPDY/$gcyc/$COMPONENT}
else
    export COMOUT="$ROTDIR/$CDUMP.$PDY/$cyc/$COMPONENT"
    export COMIN_OBS="$DMPDIR/$CDATE/$CDUMP"
    export COMIN_GES_OBS="$DMPDIR/$GDATE/$GDUMP"
fi
```
The `else` blocks do not contain `COMPONENT` for `COMIN_OBS` and `COMIN_GES_OBS`.  This is incorrect.

Setting `GDUMP` and `GDATE` in `config.base` probably isn't the long term solution we want.  The above tests are just that - tests.

**Comment by aerorahul** [comment]
I don't understand why this is in a config file to begin with.
A config file should contain static information, not something that changes with each cycle.
That is for the job.

Screen Shot 2022-07-15 at 3 07 14 PM

**Comment by RussTreadon-NOAA** [comment]
> I don't understand why this is in a config file to begin with. A config file should contain static information, not something that changes with each cycle. That is for the job.
> 
> Screen Shot 2022-07-15 at 3 07 14 PM

I commented out COM directory settings in `config.base` on Orion.   A rerun of 2021122100 `gdassfcanl` is OK.   A rerun of 2021122100  `gdasanal` does not run correctly.    `COMIN_OBS` and `COMIN_OBS_GES` are set incorrectly.  For example,  `COMIN_OBS` winds up as
```
00 + compath.py prod/obsproc/

*************************************************************
** FATAL ERROR: [compath] Unable to find /gpfs/?p2, are you on WCOSS?
** ABNORMAL EXIT at Fri Jul 15 14:16:29 CDT 2022 on Orion-01-20
*************************************************************


Currently Loaded Modules:
  1) gempak/7.5.1     16) sigio/2.3.2        31) hpc-miniconda3/4.6.14
  2) ncl/6.6.2        17) bacio/2.4.1        32) ufswm/1.0.0
  3) prod_util/1.2.2  18) nemsio/2.5.2       33) met/9.1
  4) grib_util/1.2.2  19) w3emc/2.7.3        34) intel/2018.4
  5) crtm/2.3.0       20) w3nco/2.4.1        35) hdf5/1.10.5
  6) jasper/2.0.25    21) ncdiag/1.0.0       36) netcdf/4.7.2
  7) zlib/1.2.11      22) wgrib2/2.0.8       37) udunits/2.2.26
  8) png/1.6.35       23) contrib/0.1        38) antlr/2.7.7
  9) pio/2.5.2        24) rocoto/1.3.3       39) nco/4.8.1
 10) esmf/8.2.1b04    25) munge/0.5.13       40) wgrib/2.0.8
 11) fms/2021.03      26) slurm/19.05.3-2    41) metplus/3.1
 12) g2/3.4.2         27) cdo/1.9.5          42) module_base.orion
 13) g2tmpl/1.10.0    28) hpc/1.2.0          43) impi/2018.4
 14) ip/3.3.3         29) hpc-intel/2018.4   44) hpc-impi/2018.4
 15) sp/2.3.3         30) miniconda3/4.6.14




/work/noaa/stmp/rtreadon/RUNDIRS/pr905/2021122100/gdas/anal.337867
total 1
-rwxr--r-- 1 rtreadon stmp 328 Jul 15 14:16 PDY
-rw-r--r-- 1 rtreadon stmp   0 Jul 15 14:16 ncepdate
cat: OUTPUT.337867: No such file or directory
0 + export COMIN_OBS=/gdas.20211221/00/atmos
0 + COMIN_OBS=/gdas.20211221/00/atmos
```

The leading part of the `COMIN_OBS` path is incorrect.   No bufr dump files are read by the gsi.  `compath.py` does not work on Orion.   It seems to be hardwired for WCOSS2.  Hence setting `COMIN_GES` in `config.base` ... not the optimal solution.

**Comment by WalterKolczynski-NOAA** [comment]
So where should these two be pointing in development? Setting them is a relatively new addition that I added because of the compath.py issue, and then if I used the old default (in ROTDIR), it wouldn't find any obs.

**Comment by WalterKolczynski-NOAA** [comment]
I _think_ I have this addressed in work for #761. Will need confirmation when it's ready.

**Pull request merged**